### PR TITLE
Modularize FTL modules/headers

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

Since the advent of our Faster Than Light (pihole-*FTL*) engine, it somewhat organically grew and quickly became a complex daemon. However, instead of properly fractionating the code into clearly separated modules, we instead got two very large files `FTL.h` and `routines.h` which contained prototypes for everything.

This PR restructures our *FTL* engine by separating the declarations and functions into separated modules, each consisting of a .c and the corresponding .h files. These two files build modules, each corresponding to a clear piece of functionality. They can be largely maintained and developed separately from the other modules, and partially even reused in entirely different projects.

Summary of the changes in this pull request:
1. Move definitions of global variables from `FTL.h` into the corresponding module headers where this is practical.
2. Move *all* prototypes into the corresponding module headers. This allows us to remove the old and very large `routines.h` altogether.
3. Move database related modules into a common subdirectory `database/`
4. Split the old `database.c` file into two smaller files (`database/common.c` and `database/query-table.c`).
5. Rename some files and functions for the sake of consistency.
6. Move some files from their old module into a new one if they fit there better. An example is moving the function that calls `stat()` to obtain the physical size of the database from `database.c` into `files.c` where similar functions already exist.

### CODE REVIEW

File: src/database/network-table.c [hunk 1]
```diff
@@ -370,11 +375,11 @@ static char* getMACVendor(const char* hwaddr)
...
 	}
 
-	sqlite3 *macdb;
-	int rc = sqlite3_open_v2(FTLfiles.macvendordb, &macdb, SQLITE_OPEN_READONLY, NULL);
-	if( rc != SQLITE_OK ){
-		logg("getMACVendor(%s) - SQL error (%i): %s", hwaddr, rc, sqlite3_errmsg(macdb));
-		sqlite3_close(macdb);
+	sqlite3 *macvendor_db = NULL;
+	int rc = sqlite3_open_v2(FTLfiles.macvendor_db, &macvendor_db, SQLITE_OPEN_READONLY, NULL);
+	if( rc != SQLITE_OK  ){
```
Comment (Mcat12) [inline src/database/network-table.c]: Extra space added here?

File: src/files.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+*  File prototypes
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+#ifndef FILE_H
+#define FILE_H
+
+int countlines(const char* fname);
+int countlineswith(const char* str, const char* fname);
+void check_blocking_status(void);
```
Comment (Mcat12) [inline src/files.h]: Should this function be in this header?

File: src/timers.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+#include "FTL.h"
+#include "timers.h"
+#include "memory.h"
+#include "log.h"
+
+struct timeval t0[NUMTIMERS];
```
Comment (Mcat12) [inline src/timers.c]: Can `NUMTIMERS` and the corresponding timers be defined in `timers.h`?

File: .gitignore [hunk 4]
```diff
@@ -17,7 +17,7 @@ version*
 # Test-generated files
 /pihole.log
 /pihole-FTL.conf
-/pihole-FTL.db
+/*.db
 /pihole-FTL.log
 
 # MAC->Vendor database files
```
Comment (Mcat12) [inline .gitignore]: The macvendor script has moved, so these lines should be updated.

File: src/files.c [hunk 5]
```diff
@@ -155,3 +159,19 @@ bool chmod_file(const char *filename, const mode_t mode)
...
+
+double get_FTL_db_filesize(void)
+{
+	struct stat st;
+	if(stat(FTLfiles.FTL_db, &st) != 0)
+	{
+		// stat() failed (maybe the DB file does not exist?)
+		return 0;
+	}
+	return 1e-6*st.st_size;
```
[the remark this answers is not in the document]
Comment (DL6ER) [reply src/files.c]: We also call the routine in `src/database/query-table.c`:
```c
logg("Notice: Database size is %.2f MB, deleted %i rows", get_FTL_db_filesize(), affected);
```
However, we can (and probably should) make the `1e-6` calculation there.

File: src/files.h (hunk 2 above)
Comment (DL6ER) [reply src/files.h]: Yes, the corresponding function is in `files.c`. However, as we read the blocking status from `setupVars.conf`, I will move it to the corresponding module.

File: src/config.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+#ifndef CONFIG_H
+#define CONFIG_H
+
+void getLogFilePath(void);
+void read_FTLconf(void);
+void get_privacy_level(FILE *fp);
+void get_blocking_mode(FILE *fp);
+void read_debuging_settings(FILE *fp);
+
+extern ConfigStruct config;
```
[the remark this answers is not in the document]
Comment (DL6ER) [reply src/config.h]: Done

File: src/setupVars.h [hunk 7]
```diff
@@ -18,5 +18,8 @@ bool insetupVarsArray(const char * str);
 bool getSetupVarsBool(const char * input) __attribute__((pure));
 char* find_equals(const char* s) __attribute__((pure));
 void trim_whitespace(char *string);
+void check_blocking_status(void);
+
+extern unsigned char blockingstatus;
```
Comment (Mcat12) [inline src/setupVars.h]: The `BLOCKING_*` status enums should also be moved here

File: src/timers.h [hunk 8]
```diff
@@ -10,6 +10,12 @@
 #ifndef TIMERS_H
 #define TIMERS_H
 
+// Timer enumeration
+enum { DATABASE_WRITE_TIMER, EXIT_TIMER, GC_TIMER, LISTS_TIMER, REGEX_TIMER, ARP_TIMER, LAST_TIMER };
+
+// Prepare timers, used mainly for debugging purposes
+#define NUMTIMERS LAST_TIMER
```
Comment (Mcat12) [inline src/timers.h]: Not sure that this comment is relevant for this line. Is this a copy-paste error?

File: src/dnsmasq_interface.c [hunk 9]
```diff
@@ -616,8 +635,8 @@ static void detect_blocked_IP(const unsigned short flags, const char* answer, co
 
 		// Update status
 		query_externally_blocked(queryID, QUERY_EXTERNAL_BLOCKED_IP);
+		return;
```
Comment (Mcat12) [inline src/dnsmasq_interface.c]: Why are these returns added? They are unnecessary as the function exits here anyways.

**Comment by Mcat12** [comment]
Note that the codefactor issues are ignored, but the GitHub check is not updated.

**Pull request merged**